### ISSUE TITLE
Determine instanceID policy

### ISSUE THREAD

**Issue by asutula** [textileio/go-threads#280]
JSON mode originally had a strict uuid policy where non-json mode allowed anything. For now we're allowing anything, but let's discuss what policy should be applied. Some notes from slack convo:

@sanderpick: For example, ObjectIDs (in mongo) have a prefixed time component

@jsign: Maybe we can add an extra flag for ID validations... that might help for flexibility, and as a free solution to tests too

**Comment by asutula** [comment]
@sanderpick and @jsign I'd like to work on this. Seems easy other than just making the decision about id format. Either of you want to weigh in?

My only thoughts are that we should be able to provide a simple api for generating ids on the client side for cases where the developer wants to set their own id before creating the instance in the db. This would need to be the case no matter which language/platform it is.

UUID sounds good enough to me, but I think you guys may have better perspective on why adding additional information to the id could be useful.

**Comment by jsign** [comment]
@sanderpick  mentioned that the ID might have a prefixed time component. 

That seems at least interesting for having some total-order by just ordering IDs. Of course, using timestamps from clients might be correct under particular assumptions, which the client should know about. Of course, it could put that timestamp as an independent attribute... but I guess that would mean creating a separate index, and so on.

No strings opinions from my part, just extending what I understood by that idea.

**Comment by sanderpick** [comment]
Some more info here: https://docs.mongodb.com/manual/core/document/#the-id-field

tl;dr, Mongo allows the user the use whatever they want, including UUID. However, if the user does not supply an value, an ObejctID is created. ObjectIDs are 12 bytes, prefixed with a timestamp for lexicographical sorting: https://docs.mongodb.com/manual/reference/method/ObjectId/#description

My vote is for something similar, but since our storage format is JSON (vs. BSON), we'd have to be clever about comparing the IDs for uniqueness.

**Comment by jsign** [comment]
Mongo can have some reasonable assumption about synchronized clocks to make lexicographical sorting a meaningful thing.

On a fully p2p use of _DB_ I think that assumption doesn't hold, since a peer with a weird clock will break any kind of app logic that relies on that kind of sorting. Sounds like that feature would be good in some special cases in which _DB_ is used in some kind of centralized-way.

We could bake-in the timestamp-prefix as the default, but that would mean putting some big-warning signs on docs to let people think about the safeness of using that for sorting depending on the context.

**Comment by asutula** [comment]
Currently with our UUID or any value ids, when you query data from the db with no order by, how is the data ordered? Random? Deterministic?

**Comment by sanderpick** [comment]
@jsign Mongo does let clients specify their own `_id` on the client side. So, I think the same limitation around distributed machines applies there. I guess the assumption is that it would be better than nothing. 

In any case, I don't love having to come up with out our `ID` objects. As @carsonfarmer mentioned, https://github.com/ulid/spec sounds like a good alternative to UUID, which already has a sorting component.

+1 to letting the client be free, but falling back to some standard.

**Comment by asutula** [comment]
Yea having some default deterministic sort order is nice, and actually useful/meaningful in certain situations (which we can document). Agree with the better than nothing thing.

**Comment by jsign** [comment]
> Currently with our UUID or any value ids, when you query data from the db with no order by, how is the data ordered? Random? Deterministic?

From a global perspective, I'd say the result aren't ordered. For a local client, I think they have ordering considering the order in which the _create_ event was received (which is usually a meaningless order for an app).

> @jsign Mongo does let clients specify their own `_id` on the client side. So, I think the same limitation around distributed machines applies there. I guess the assumption is that it would be better than nothing.

What I mean is that setting the _default_ to have a timestamp-prefix might bake assumptions that would make sense in Mongo but not in _DB_.

The _distributed setup_ in Mongo are _clients_ that are owned by the same company running in a datacenter with some common clock synchronization protocol.
The _distributed setup_ in _DB_ are _clients_ owned by different people, in different countries, and without _assumptions_ of running clock sync protocols. 

I wouldn't see both cases as having the same assumptions. So I'd guess something different, and this point might be involved in the decision. But just my guess that might be completely wrong.

> Yea having some default deterministic sort order is nice, and actually useful/meaningful in certain situations (which we can document). Agree with the better than nothing thing.

Well, actually you can pretend to have something that might be useful. But if you can't use it safely most of the times, I'm not sure _better than nothing_ is good advice for a default setup.

So, I think we all agree that it should be _free_... but maybe the _default_ decision is what it seems I'm the only one disagreeing!

**Comment by jsign** [comment]
We can lean towards using something, say _ulid_ (which I don't know much about) and then we can always reconsider things later!

**Comment by carsonfarmer** [comment]
Another feather in ulid's hat is that it is the same number of bytes as a uuid... so a future switch should be compatible.

**Comment by carsonfarmer** [comment]
Just to follow up here, is there anything _mission critical_ going on with ID sort order? Or is it just so that we have a deterministic ordering. I can't think of a specific use-case where ID ordering is super critical? So it seems like using ulids and just documenting that order is defined by the time at which a _given peer_ created the instance is reasonable enough?

**Comment by sanderpick** [comment]
@jsign Fair enough about the assumptions being different. Technically, you could run a MongoDB Cluster with members in different data centers, running on different hardware, etc. Most people will not do that, but it is possible. 

I think a deterministic time-based default sort order is a fairly well known use-case, no? Anyway, the solution if we went with something like ulid is to just make some good docs in the code.

**Comment by sanderpick** [comment]
To spell out the use-case: "Give me back the stuff in the order it happened. I'm the only DB writer so I don't care about different clocks".

**Comment by sanderpick** [comment]
I'm also totally fine with saying to the user, "just make your own time field and add an index". However, in this case, we'd also have to inform the user of clock issues. 

In the end, I agree, the time component shouldn't be the main reason to use ulid. Admittedly, I'm partially driven by how they look... base32 IDs would look good next to our thread IDs, peer IDs, and CIDs :)

**Comment by jsign** [comment]
And why not say _make your own id: an integer and +1 on insert_ since you're a single writer?

**Comment by sanderpick** [comment]
Totally, could work.

**Comment by carsonfarmer** [comment]
But that might lead to collisions, which _is_ something we can/should avoid with uuid, ulid, etc.

**Comment by jsign** [comment]
> But that might lead to collisions, which _is_ something we can/should avoid with uuid, ulid, etc.

In single writers?

**Comment by carsonfarmer** [comment]
Ah, missed that part. Well, anyway single writers _can_ use their own ids. So the tl;dr here seems to be... for those who want a default, ulids look nice, are unique, and have a (rough) time sorting which is, while perhaps not 'safe' to rely on, usefully sortable.

**Comment by asutula** [comment]
Ok sounds good. Next question...

There was mention of allowing users to specify `ID` as not only `string`, but `integer` and `number` as well. Do we want to move forward with that? If so, it seems that we'd remove our existing `InstanceID` type (`string` alias).

**Comment by asutula** [comment]
It certainly complicates things everywhere to support more than just `string`

**Comment by sanderpick** [comment]
Hmm, yeah. On second thought, `string` seems like a pretty reasonable requirement.

**Comment by asutula** [comment]
Yea, we actually have quite a few assumptions based on `string`, breaking those assumptions is possible but probably opens up a class of potential bugs we'll have to deal with.

**Comment by asutula** [comment]
Plus, we can be more strict now with `string` and if we really see a need in the future to open it up, we can. Much harder to open it now and try to restrict later.

**Comment by jsign** [comment]
To clarify, when I mentioned the integer solution, as I mentioned in the edit, it means dealing with still keeping it in the `string` format. You can have a lexicographical order of ints as strings. 
I see that decision as a client problem, not something `go-thread` should care about.

**Comment by sanderpick** [comment]
👍

**Comment by jsign** [comment]
Adding another question.

This discussion means that getting things in order is optional, or done by default in all operations?

**Comment by carsonfarmer** [comment]
My interpretation of this discussion is that, getting things in order by id will happen if the id is lexicographically sortable (which ulids are), and otherwise, the order is arbitrary. Though if a sort is explicitly provided via the existing api, then _that_ order will be used (and likely ties in that ordering will end up being lexicographically sorted as in the default case).
